design/multi-factor-admin-and-step-up.md

```diff
@@ -0,0 +1,613 @@
+  |- OTP                           [x] enabled  [ ] disabled
+  |- Other                         [x] enabled  [ ] disabled
+```
+
+##### Configurable flows
+
+Configurable flows work as is currently the case in Keycloak with the possibility
+to add new authentication flows and either replace the default flow in _Bindings_
+or in the _Client > [client] > Authentication Flow Overrides_. Here the normal configuration of
+an authentication factor is to add (or remove) authenticator executions. However, within an
+authentication execution's configuration, it is possible to set a **disable** flag.
+This is for when an admin doesn't want to keep an execution active, but doesn't want it's
+configuration to be deleted. An **optional** flag is present in the configuration of the
+authentication factor, meaning that it will only be required for users which have at least
+one credential set for an execution that is within that authentication factor.
+
+For UX purposes, the GUI shows directly in the flow when an authentication factor is set
+to optional, or if an authentication execution is disabled.
+
+#### Bindings
+
+No changes are necessary here.
+
+#### Required Actions
+
+Current required actions are a bit too limited to support multi-tokens. While it is currently
+possible to require a user to, for example, register an OTP device, it must be possible for the
+administrator to require "register a valid 1st factor" and "register a valid 2nd factor".
+
+To do this, it must be possible to register a required action multiple times, and to configure
+required actions. It must also be possible to remove required actions from the list. The code
+of the required action should define if it can be registered multiple times, and how it can
+be configured. However, all required actions should be able to be removed, and removing a
+required action should also remove it from all users for which it is currently required.
+
+The purpose of this is to be able to have a "register credential" required action in which
+authentication categories can be set. For example, if an admin were to configure a 2nd factor
+with OTP and Fido2, he could then set a "register credential" with the OTP and Fido2
+authentication categories. A user, when confronted with that required action, would only have
+to register one of the two.
+
+
+#### Policies
+
+The password policy for the realm should remain, but the OTP policy should disappear, as a user
+could have several alternative OTP devices with different settings. As such, the information
+about the OTP settings should probably move to information associated with the credential.
+
+### Authentication screens for the user
+
+When the user is redirected towards Keycloak, unless he's using some system that
+bypasses the login actions (e.g. session cookie, kerberos,...) he will arrive at a
+login prompt that corresponds to his authentication category.
+
+The rule for selecting which credential is chosen for a user is
```

Comment:
Authenticators (authentication executions) is what displays forms. That needs to be made obvious in this section. I'm also having a hard time to decipher the logic here, both the bullets and the pseudo code.